In [69]:
import matplotlib.pyplot as plt

import PSO
import optproblems.cec2005
import math
import pandas

# Additional test on the 24 functions

Set AdditionalTest to True if tou want to run those tests.

The main function is to check that the PSO algorithm works on every problem of the CEC 2005 benchmarks and have a first look at his performance

In [70]:
AdditionalTest = False

Creation of benchmark and bounds lists used

In [71]:
if AdditionalTest:
    DimensionList = [2,10,30,50]
    BenchmarkList = []
    BoundsList = []
    TestProblemList = [optproblems.cec2005.F1,optproblems.cec2005.F2,optproblems.cec2005.F3,optproblems.cec2005.F4,optproblems.cec2005.F5,optproblems.cec2005.F6,optproblems.cec2005.F7,optproblems.cec2005.F8,optproblems.cec2005.F9,optproblems.cec2005.F10,optproblems.cec2005.F11,optproblems.cec2005.F12,optproblems.cec2005.F13,optproblems.cec2005.F14,optproblems.cec2005.F15,optproblems.cec2005.F16,optproblems.cec2005.F17,optproblems.cec2005.F18,optproblems.cec2005.F19,optproblems.cec2005.F20,optproblems.cec2005.F21,optproblems.cec2005.F22,optproblems.cec2005.F23,optproblems.cec2005.F24,optproblems.cec2005.F25]

    for dimension in DimensionList:
        i = 0
        for testproblem in TestProblemList:
            i += 1
            BenchmarkList.append(testproblem)
            if(i < 7 or i == 14):
                BoundsList.append([-100,100])
            elif(i == 7):
                BoundsList.append([0, 600])
            elif(i == 8):
                BoundsList.append([-32, 32])
            elif (i == 25):
                BoundsList.append([2, 5])
            elif(i == 11):
                BoundsList.append([-0.5, 0.5])
            elif(i == 12):
                BoundsList.append([-math.pi,math.pi])
            elif(i == 13):
                BoundsList.append([-3, 1])
            elif(i > 13 or i == 9 or i == 10):
                BoundsList.append([-5, 5])

Setting of the hyperparameters

In [72]:
if AdditionalTest:
    #HyperParameters
    swarmsize = 300
    α = 0.4 #how much of the original velocity is retained
    β = 0.6 #how much of the personal best is mixed in.
    γ = 0.6 #how much of the informants’ best is mixed in.
    δ = 0 #how much of the global best is mixed in. δ is often set to 0 in modern implementations
    e = 1 #e how fast the particle moves. Most commonly, e is set to 1.
    numberofiterations = 150
    precisionwanted = 0.1
    numberofinformantsperparticle = 6 #Most commonly set to 6

Creation of the PSO alogrithms

In [73]:
if AdditionalTest:
    #Create algorithms

    PSOList = []
    for i in range(len(BenchmarkList)):
        if(i == 6 or i == 24 or i == 31 or i == 49):
            bounded = False
        else:
            bounded = True
        if i < 25:
            PSOList.append(PSO.PSO(BenchmarkList[i], DimensionList[0], BoundsList[i], bounded,swarmsize, numberofinformantsperparticle, α, β, γ, δ, e))
        else:
            PSOList.append(PSO.PSO(BenchmarkList[i], DimensionList[1], BoundsList[i], bounded,swarmsize, numberofinformantsperparticle, α, β, γ, δ, e))

Recording of the results

In [74]:
if AdditionalTest:
    #Try each test function of optproblems.cec2005 with dimension 2 10 and 50 and appropriate bounds
    Results = []
    Resutlsbool = []
    Times = []
    for i in PSOList:
        result = i.run(numberofiterations, precisionwanted)
        Results.append(result[0])
        Times.append(result[1])
        Resutlsbool.append(result[2])

    Percentagesuccess = (sum(Resutlsbool)/len(PSOList))*100
    print("Configuration: Max number of iterations: ",numberofiterations," Precision wanted: ",precisionwanted," Swarm size: ",swarmsize," Number of informants per particle: ",numberofinformantsperparticle," α: ",α," β: ",β," γ: ",γ," δ: ",δ," e: ",e)
    print("Success rate on 24 tests functions for 10 and 30 dimensions: ",Percentagesuccess)

# Hyperparameter exploration

Set Hyperparameter_exploration_swarmsize to True if you want to experiment on swarm size

In [75]:
    Hyperparameter_exploration_swarmsize = False

In [76]:
if(Hyperparameter_exploration_swarmsize):
    SwarmsizeList = [10,25,35,50,80,100,300,500,800]
    numberofinformantsperparticle = 6
    α = 0.4
    β = 0.6
    γ = 0.6
    δ = 0
    e = 1
    numberofiterations = 140
    precisionwanted = 0.1

    AverageOverRange = 5

    ListFitnessResult = []
    ListTimeProcess = []
    for swarmsize in SwarmsizeList:
        fitness = 0.0
        time = 0.0
        for i in range(AverageOverRange):
            F = PSO.PSO(optproblems.cec2005.F17, 10, [-5,5], True, swarmsize, numberofinformantsperparticle, α, β, γ, δ, e)
            result = F.run(numberofiterations, precisionwanted)
            fitness += result[0].fitness
            time += result[1]
        fitness = fitness/AverageOverRange
        time = time/AverageOverRange
        ListFitnessResult.append(fitness)
        ListTimeProcess.append(time)

    plt.plot(SwarmsizeList,ListTimeProcess)
    plt.plot(SwarmsizeList, ListFitnessResult)
    plt.axhline(y=optproblems.cec2005.F17.bias)
    plt.legend(["Process time", "Fitness reached", "Best fitness for this problem"], loc ="lower right")
    plt.title("Study of process time and best result, depending on swarm size for PSO, \n hyperparameters are: Problem: F17, Numberofinformantsperparticle: 6, \n α: 0.4, β: 0.6, γ: 0.6, δ: 0, e: 1, maxnumberofiterations: 140, precisionwantedtostop: 0.1")
    plt.show()

    plt.plot(SwarmsizeList, ListFitnessResult)
    plt.axhline(y=optproblems.cec2005.F17.bias)
    plt.legend(["Fitness reached", "Best fitness for this problem"], loc ="lower right")
    plt.title("Study of best result, depending on swarm size for PSO, \n hyperparameters are: Problem: F17, Number of informants per particle: 6, \n α: 0.4, β: 0.6, γ: 0.6, δ: 0, e: 1, maxnumberofiterations: 300, precisionwantedtostop: 0.1")
    plt.show()

    plt.plot(SwarmsizeList,ListTimeProcess)
    plt.title("Study of process time , depending on swarm size for PSO, \n hyperparameters are: Problem: F17, Number of informants per particle: 6, \n α: 0.4, β: 0.6, γ: 0.6, δ: 0, e: 1, maxnumberofiterations: 140, precisionwantedtostop: 0.1")
    plt.show()


Resulting graphs:

![](Graph1.png)


![](Graph2.png)


![](Graph3.png)

Conclusion: For this problem augmenting the swarm size over 500 doesn't improve the result or only slightly while augmenting process time in a linear way. A good tradeoff between performance and process time would be a swarm size of around 80 where whe observe a pick in the fitness reached. Same experiment realised with F8 F4 or F2 shows that for some problems the optimal swarm size is lower (around 35)

Set Hyperparameter_exploration_inertia to True if you want to experiment on inertia hyperparameters

In [77]:
Hyperparameter_exploration_inertia = False

In [78]:
if(Hyperparameter_exploration_inertia):
    swarmsize = 500
    numberofinformantsperparticle = 6
    αList = [0.2,0.4,0.5,0.6]
    βList = [0.4,0.5,0.6,0.8]
    γList = [0.4,0.5,0.6,0.8]
    δ = 0.0
    e = 1
    numberofiterations = 140
    precisionwanted = 0.1

    AverageOverRange = 5

    Initialise = True
    NumberIteration=0.0
    for α in αList:
        for β in βList:
            for γ in γList:
                fitness = 0.0
                for i in range(AverageOverRange):
                    NumberIteration+=1.0
                    F = PSO.PSO(optproblems.cec2005.F17, 10, [-5,5], True, swarmsize, numberofinformantsperparticle, α, β, γ, δ, e)
                    result = F.run(numberofiterations, precisionwanted)
                    fitness += result[0].fitness
                fitness = fitness/AverageOverRange
                if(Initialise):
                    Initialise = False
                    bestfitness = fitness
                elif(bestfitness > fitness):
                    bestfitness = fitness
                    bestconfig = [α, β, γ]
                print(f"Progress: {(NumberIteration/320.0)*100.0}%")

    print("Best config parameter α: ",bestconfig[0]," β: ",bestconfig[1]," γ: ",bestconfig[2])

Best config parameter found: α:  0.4  β:  0.6  γ:  0.6

# Conlusion of exploration

More research into the optimal parameters for each different problem would be appropriate as the optimal swarm size for example can be different from a problem to another, time constraints didn't let us the possibility to do so.



# Final test for comparison

 Set Write to True only if you dont have a dfPSO.pkl file

# Test Functions
- F2: Shifted Schwefel’s Problem 1.2
    - Uni-modal without noise
- F4: Shifted Schwefel’s Problem 1.2 with Noise in Fitness
    - Uni-modal with noise
- F8: Shifted Rotated Ackley’s Function with Global Optimum on Bounds
    - Single function
    - Global optimum on bounds
    - Multi-modal
- F13: Shifted Expanded Griewank’s plus Rosenbrock’s Function (F8F2)
    - Expanded Function
    - Multi-modal
- F17: Rotated Version of Hybrid Composition Function with Noise in Fitness
    - Gaussian noise in fitness
    - Multi-modal
    - Huge number of local optima
    - In-deterministic
- F24: Rotated Hybrid Composition Function
    - Composite function
    - Huge number of local optima
    - Multi-modal

In [79]:
Write = False

In [80]:
#Swarm size as been lmited to 350 to reduce computation time
swarmsize = 350
numberofinformantsperparticle = 6
α = 0.4
β = 0.6
γ = 0.6
δ = 0
e = 1
numberofiterationsList = [250,500,1000]
DimensionList = [10,30,50]
precisionwanted = 0.1
benchmarks_and_bounds = [[optproblems.cec2005.F2, [-100,100],"F2"], [optproblems.cec2005.F4,[-100,100],"F4"], [optproblems.cec2005.F8, [-32,32],"F8"], [ optproblems.cec2005.F13, [-3,1],"F13"], [optproblems.cec2005.F17, [-5,5],"F17"], [optproblems.cec2005.F24, [-5,5],"F24"]]
AverageOverRange = 5

list_for_df = []

if(Write):
    NumberIteration=0.0
    for i in range(len(benchmarks_and_bounds)):
        print(f"\t\tFunction: {benchmarks_and_bounds[i][2]}")
        for numberofiterations in numberofiterationsList:
            print(f"\t\tIterations: {numberofiterations}")
            for dimension in DimensionList:
                print(f"\t\tDimension: {dimension}")
                averagefitness = 0
                time = 0
                for a in range(AverageOverRange):
                    NumberIteration+=1.0
                    F = PSO.PSO(benchmarks_and_bounds[i][0], dimension, benchmarks_and_bounds[i][1], True, swarmsize, numberofinformantsperparticle, α, β, γ, δ, e)
                    result = F.run(numberofiterations, precisionwanted)
                    averagefitness += result[0].fitness
                    fitness = result[0].fitness
                    time += result[1]
                    print(f"Progress: {(NumberIteration/270)*100.0}%")
                    if(a == 0):
                        bestfitness = fitness
                    elif(fitness<bestfitness):
                        bestfitness = fitness
                averagefitness = averagefitness/5.0
                time = time/5.0
                result_and_configuration = [dimension,benchmarks_and_bounds[i][2],benchmarks_and_bounds[i][0].bias,averagefitness,abs(benchmarks_and_bounds[i][0].bias-averagefitness),time,bestfitness,abs(benchmarks_and_bounds[i][0].bias-bestfitness),numberofiterations,dimension,swarmsize,numberofinformantsperparticle,α, β, γ, δ, e,AverageOverRange]
                list_for_df.append(result_and_configuration)

    labels = ["Dimension", "TestFunction", "OptimalMinimum", "AverageBestFitness", "AverageBestFitnessDiffFromOpt","Time","BestBestFitness","BestBestFitnessDiffFromOpt", "Iterations", "Dimensions", "SwarmSize", "NumbeInformantsPerParticle", "α", "β", "γ", "δ", "e","AverageOverRange"]
    dfPSO = pandas.DataFrame(data=list_for_df, columns=labels)
    dfPSO.to_pickle("dfPSO.pkl")

else:
    dfPSO = pandas.read_pickle('dfPSO.pkl')

In [81]:
dfPSO

Dimension TestFunction  OptimalMinimum  AverageBestFitness  \
0          10           F2          -450.0         1349.277489   
1          30           F2          -450.0       176875.627666   
2          50           F2          -450.0       478391.106707   
3          10           F2          -450.0         1548.915918   
4          30           F2          -450.0       180368.052150   
5          50           F2          -450.0       535087.424428   
6          10           F2          -450.0         1727.416162   
7          30           F2          -450.0       180598.548843   
8          50           F2          -450.0       579107.952665   
9          10           F4          -450.0         3318.031928   
10         30           F4          -450.0       150637.338937   
11         50           F4          -450.0       634679.937289   
12         10           F4          -450.0         2400.700671   
13         30           F4          -450.0       181172.293821   
14         50           F4          -450.0       571940.182952   
15         10           F4          -450.0         3014.000187   
16         30           F4          -450.0       207011.104985   
17         50           F4          -450.0       626126.542820   
18         10           F8          -140.0         -119.240789   
19         30           F8          -140.0         -118.948455   
20         50           F8          -140.0         -118.738049   
21         10           F8          -140.0         -118.510784   
22         30           F8          -140.0         -118.983226   
23         50           F8          -140.0         -118.823952   
24         10           F8          -140.0         -119.414019   
25         30           F8          -140.0         -118.954023   
26         50           F8          -140.0         -118.932675   
27         10          F13          -130.0         -128.767771   
28         30          F13          -130.0         -115.977292   
29         50          F13          -130.0         -103.296913   
30         10          F13          -130.0         -128.212716   
31         30          F13          -130.0         -116.471890   
32         50          F13          -130.0         -101.337794   
33         10          F13          -130.0         -128.626699   
34         30          F13          -130.0         -114.760453   
35         50          F13          -130.0         -101.877134   
36         10          F17           120.0          310.983656   
37         30          F17           120.0          687.799100   
38         50          F17           120.0          852.254970   
39         10          F17           120.0          284.870831   
40         30          F17           120.0          691.785661   
41         50          F17           120.0          954.503277   
42         10          F17           120.0          276.310133   
43         30          F17           120.0          725.252419   
44         50          F17           120.0          807.088982   
45         10          F24           260.0          962.758387   
46         30          F24           260.0         1562.086510   
47         50          F24           260.0         1597.579555   
48         10          F24           260.0         1174.564756   
49         30          F24           260.0         1560.174719   
50         50          F24           260.0         1596.628766   
51         10          F24           260.0         1152.988019   
52         30          F24           260.0         1559.697927   
53         50          F24           260.0         1597.926843   

    AverageBestFitnessDiffFromOpt        Time  BestBestFitness  \
0                     1799.277489    0.785878        51.426083   
1                   177325.627666    2.680550    156430.307241   
2                   478841.106707    4.293332    229154.168523   
3                     1998.915918    2.005395       916.847329   
4                   180818.052150    4.4314

In [82]:
dfPSO['BestBestFitnessDiffFromOpt'] = dfPSO['BestBestFitnessDiffFromOpt'].abs()

# F2

- Best Diff: 501.43 -> 250 iterations / 10 Dimension / 100 swarm size / 6 informants per particle / α 0.4 / β 0.6 / γ 0.6 / δ 0 / e 1
    - Error: 501.43 / 450 ->  111.43%
    - Time: 0.79
- Avr. Diff: 1799.28 -> 250 iterations / 10 Dimension / 100 swarm size / 6 informants per particle / α 0.4 / β 0.6 / γ 0.6 / δ 0 / e 1
    - Error: 1799.28 / 450 ->  399.84%
    - Time: 0.79


In [83]:
df_F2 = dfPSO.query("TestFunction == 'F2'")
df_F2[df_F2.BestBestFitnessDiffFromOpt == df_F2.BestBestFitnessDiffFromOpt.min()]

Dimension TestFunction  OptimalMinimum  AverageBestFitness  \
0         10           F2          -450.0         1349.277489   

   AverageBestFitnessDiffFromOpt      Time  BestBestFitness  \
0                    1799.277489  0.785878        51.426083   

   BestBestFitnessDiffFromOpt  Iterations  Dimensions  SwarmSize  \
0                  501.426083         250          10        100   

   NumbeInformantsPerParticle    α    β    γ  δ  e  AverageOverRange  
0                           6  0.4  0.6  0.6  0  1                 5

In [84]:
df_F2[df_F2.AverageBestFitnessDiffFromOpt == df_F2.AverageBestFitnessDiffFromOpt.min()]

Dimension TestFunction  OptimalMinimum  AverageBestFitness  \
0         10           F2          -450.0         1349.277489   

   AverageBestFitnessDiffFromOpt      Time  BestBestFitness  \
0                    1799.277489  0.785878        51.426083   

   BestBestFitnessDiffFromOpt  Iterations  Dimensions  SwarmSize  \
0                  501.426083         250          10        100   

   NumbeInformantsPerParticle    α    β    γ  δ  e  AverageOverRange  
0                           6  0.4  0.6  0.6  0  1                 5

# F4

- Best Diff: 1124.51 -> 250 iterations / 10 Dimension / 100 swarm size / 6 informants per particle / α 0.4 / β 0.6 / γ 0.6 / δ 0 / e 1
    - Error: 1124.51 / 450 ->  249.89%
    - Time: 1.55
- Avr. Diff: 2850.7 -> 500 iterations / 10 Dimension / 100 swarm size / 6 informants per particle / α 0.4 / β 0.6 / γ 0.6 / δ 0 / e 1
    - Error: 2850.7 / 450 ->  633.49%
    - Time: 1.65

In [85]:
df_F4 = dfPSO.query("TestFunction == 'F4'")
df_F4[df_F4.BestBestFitnessDiffFromOpt == df_F4.BestBestFitnessDiffFromOpt.min()]

Dimension TestFunction  OptimalMinimum  AverageBestFitness  \
9         10           F4          -450.0         3318.031928   

   AverageBestFitnessDiffFromOpt      Time  BestBestFitness  \
9                    3768.031928  1.549669       674.513647   

   BestBestFitnessDiffFromOpt  Iterations  Dimensions  SwarmSize  \
9                 1124.513647         250          10        100   

   NumbeInformantsPerParticle    α    β    γ  δ  e  AverageOverRange  
9                           6  0.4  0.6  0.6  0  1                 5

In [86]:
df_F4[df_F4.AverageBestFitnessDiffFromOpt == df_F4.AverageBestFitnessDiffFromOpt.min()]

Dimension TestFunction  OptimalMinimum  AverageBestFitness  \
12         10           F4          -450.0         2400.700671   

    AverageBestFitnessDiffFromOpt      Time  BestBestFitness  \
12                    2850.700671  1.654915      1585.593229   

    BestBestFitnessDiffFromOpt  Iterations  Dimensions  SwarmSize  \
12                 2035.593229         500          10        100   

    NumbeInformantsPerParticle    α    β    γ  δ  e  AverageOverRange  
12                           6  0.4  0.6  0.6  0  1                 5

# F8
- Best Diff: 20.11 -> 250 iterations / 10 Dimension / 100 swarm size / 6 informants per particle / α 0.4 / β 0.6 / γ 0.6 / δ 0 / e 1
    - Error: 20.11 / 140 ->  14.36%
    - Time: 1.024
- Avr. Diff: 20.58 -> 1000 iterations / 10 Dimension / 100 swarm size / 6 informants per particle / α 0.4 / β 0.6 / γ 0.6 / δ 0 / e 1
    - Error: 20.58 / 140 ->  14.7%
    - Time: 3.9

In [87]:
df_F8 = dfPSO.query("TestFunction == 'F8'")
df_F8[df_F8.BestBestFitnessDiffFromOpt == df_F8.BestBestFitnessDiffFromOpt.min()]

Dimension TestFunction  OptimalMinimum  AverageBestFitness  \
18         10           F8          -140.0         -119.240789   

    AverageBestFitnessDiffFromOpt      Time  BestBestFitness  \
18                      20.759211  1.024208      -119.888388   

    BestBestFitnessDiffFromOpt  Iterations  Dimensions  SwarmSize  \
18                   20.111612         250          10        100   

    NumbeInformantsPerParticle    α    β    γ  δ  e  AverageOverRange  
18                           6  0.4  0.6  0.6  0  1                 5

In [88]:
df_F8[df_F8.AverageBestFitnessDiffFromOpt == df_F8.AverageBestFitnessDiffFromOpt.min()]

Dimension TestFunction  OptimalMinimum  AverageBestFitness  \
24         10           F8          -140.0         -119.414019   

    AverageBestFitnessDiffFromOpt      Time  BestBestFitness  \
24                      20.585981  3.913751      -119.771211   

    BestBestFitnessDiffFromOpt  Iterations  Dimensions  SwarmSize  \
24                   20.228789        1000          10        100   

    NumbeInformantsPerParticle    α    β    γ  δ  e  AverageOverRange  
24                           6  0.4  0.6  0.6  0  1                 5

# F13
- Best Diff: 0.49 -> 250 iterations / 10 Dimension / 100 swarm size / 6 informants per particle / α 0.4 / β 0.6 / γ 0.6 / δ 0 / e 1
    - Error: 0.49 / 130 ->  0.38%
    - Time : 1.2
- Avr. Diff: 1.23 -> 250 iterations / 10 Dimension / 100 swarm size / 6 informants per particle / α 0.4 / β 0.6 / γ 0.6 / δ 0 / e 1
    - Error: 1.23 / 130 ->  0.95%
    - Time : 1.2

In [89]:
df_F13 = dfPSO.query("TestFunction == 'F13'")
df_F13[df_F13.BestBestFitnessDiffFromOpt == df_F13.BestBestFitnessDiffFromOpt.min()]

Dimension TestFunction  OptimalMinimum  AverageBestFitness  \
27         10          F13          -130.0         -128.767771   

    AverageBestFitnessDiffFromOpt      Time  BestBestFitness  \
27                       1.232229  1.202112      -129.507476   

    BestBestFitnessDiffFromOpt  Iterations  Dimensions  SwarmSize  \
27                    0.492524         250          10        100   

    NumbeInformantsPerParticle    α    β    γ  δ  e  AverageOverRange  
27                           6  0.4  0.6  0.6  0  1                 5

In [90]:
df_F13[df_F13.AverageBestFitnessDiffFromOpt == df_F13.AverageBestFitnessDiffFromOpt.min()]

Dimension TestFunction  OptimalMinimum  AverageBestFitness  \
27         10          F13          -130.0         -128.767771   

    AverageBestFitnessDiffFromOpt      Time  BestBestFitness  \
27                       1.232229  1.202112      -129.507476   

    BestBestFitnessDiffFromOpt  Iterations  Dimensions  SwarmSize  \
27                    0.492524         250          10        100   

    NumbeInformantsPerParticle    α    β    γ  δ  e  AverageOverRange  
27                           6  0.4  0.6  0.6  0  1                 5

# F17
- Best Diff: 123.92 -> 500 iterations / 10 Dimension / 100 swarm size / 6 informants per particle / α 0.4 / β 0.6 / γ 0.6 / δ 0 / e 1
    - Error: 123.92 / 120 ->  103.27%
    - Time: 39.34
- Avr. Diff: 156.31 -> 1000 iterations / 10 Dimension / 100 swarm size / 6 informants per particle / α 0.4 / β 0.6 / γ 0.6 / δ 0 / e 1
    - Error: 156.31 / 120 ->  130.26%
    - Time: 57.33

In [91]:
df_F17 = dfPSO.query("TestFunction == 'F17'")
df_F17[df_F17.BestBestFitnessDiffFromOpt == df_F17.BestBestFitnessDiffFromOpt.min()]

Dimension TestFunction  OptimalMinimum  AverageBestFitness  \
39         10          F17           120.0          284.870831   

    AverageBestFitnessDiffFromOpt       Time  BestBestFitness  \
39                     164.870831  39.344609       243.922072   

    BestBestFitnessDiffFromOpt  Iterations  Dimensions  SwarmSize  \
39                  123.922072         500          10        100   

    NumbeInformantsPerParticle    α    β    γ  δ  e  AverageOverRange  
39                           6  0.4  0.6  0.6  0  1                 5

In [92]:
df_F17[df_F17.AverageBestFitnessDiffFromOpt == df_F17.AverageBestFitnessDiffFromOpt.min()]

Dimension TestFunction  OptimalMinimum  AverageBestFitness  \
42         10          F17           120.0          276.310133   

    AverageBestFitnessDiffFromOpt       Time  BestBestFitness  \
42                     156.310133  57.330861       246.529259   

    BestBestFitnessDiffFromOpt  Iterations  Dimensions  SwarmSize  \
42                  126.529259        1000          10        100   

    NumbeInformantsPerParticle    α    β    γ  δ  e  AverageOverRange  
42                           6  0.4  0.6  0.6  0  1                 5

# F24
- Best Diff: 210.92 -> 250 iterations / 10 Dimension / 100 swarm size / 6 informants per particle / α 0.4 / β 0.6 / γ 0.6 / δ 0 / e 1
    - Error: 210.92 / 260 ->  81.12%
    - Time: 12.31
- Avr. Diff: 702.76 -> 250 iterations / 10 Dimension / 100 swarm size / 6 informants per particle / α 0.4 / β 0.6 / γ 0.6 / δ 0 / e 1
    - Error: 702.76 / 260 ->  27O.29%
    - Time: 12.31

In [93]:
df_F24 = dfPSO.query("TestFunction == 'F24'")
df_F24[df_F24.BestBestFitnessDiffFromOpt == df_F24.BestBestFitnessDiffFromOpt.min()]

Dimension TestFunction  OptimalMinimum  AverageBestFitness  \
45         10          F24           260.0          962.758387   

    AverageBestFitnessDiffFromOpt     Time  BestBestFitness  \
45                     702.758387  12.3066       470.919152   

    BestBestFitnessDiffFromOpt  Iterations  Dimensions  SwarmSize  \
45                  210.919152         250          10        100   

    NumbeInformantsPerParticle    α    β    γ  δ  e  AverageOverRange  
45                           6  0.4  0.6  0.6  0  1                 5

In [94]:
df_F24[df_F24.AverageBestFitnessDiffFromOpt == df_F24.AverageBestFitnessDiffFromOpt.min()]

Dimension TestFunction  OptimalMinimum  AverageBestFitness  \
45         10          F24           260.0          962.758387   

    AverageBestFitnessDiffFromOpt     Time  BestBestFitness  \
45                     702.758387  12.3066       470.919152   

    BestBestFitnessDiffFromOpt  Iterations  Dimensions  SwarmSize  \
45                  210.919152         250          10        100   

    NumbeInformantsPerParticle    α    β    γ  δ  e  AverageOverRange  
45                           6  0.4  0.6  0.6  0  1                 5

# Conclusion

Augmenting the number of iterations seems to improve in some case the average difference from the optimal solution but the best individual solution seems to be found when running fewer iterations. This could be explained by the swarm getting more trapped into local optima when running more iterations.

The PSO algorithm seems to be working the best on the F13 problem using those parameters.

A ranking would be:
   - AverageBestFitnessDiffFromOpt:
      - F13: 0.95% time: 1.2
      - F8: 14.7%  time: 3.9
      - F17: 130.26% time: 57.33
      - F24: 270.29% time: 12.31
      - F2: 399.84% time: 0.79
      - F4: 633.49% time: 1.55


   - BestBestFitnessDiffFromOpt:
      - F13: 0.38% time: 1.2
      - F8: 14.36% time: 1.024
      - F24: 81.12% time: 39.34
      - F17: 103.27% time: 12.31
      - F2: 111.4% time: 0.79
      - F4: 249.89% time: 1.65

Problem F13 and F8 seems to work well with PSO with low error and a fast computation time


By analysing only the best results we only see the results on the 10 dimension problems as those are the less complex one, doing more analysis and comparison on the performances of the algorithm on 30 and 50 dimension would be interesting.






